platform/mbed_enum_flags.h

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2020 ARM Limited. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_SCOPED_ENUM_FLAGS_H
+#define MBED_SCOPED_ENUM_FLAGS_H
+
+#include <type_traits>
+
+#define ENUM_FLAG_OPERATOR(T, OP) \
+inline constexpr T operator OP( T lhs, T rhs) \
+{ \
+    return T (static_cast<std::underlying_type_t< T > >(lhs) OP \
+        static_cast<std::underlying_type_t< T > >(rhs)); \
+}
+
+
+#define ENUM_FLAG_OPERATOR_LHS_REFERNCE(T, OP) \
+inline constexpr T & operator OP (T & a, T b) \
+{ \
+    return a = a OP b; \
+}
+
+
+/**
+ * @brief Applies bitwise operators to a enum class defined elsewhere.
+ *
+ * @param T             The enum class typename
+ *
+ * This macro applies the bitwise negate, AND, OR, XOR operators and the
+ * corresponding assignment operators to an existing enum class. The macro
+ * uses underlying type traits to convert back and forth.
+ *
+ * Usage:
+ * @code
+ * external_enum.h:
+ *
+ * enum class SpokenLanguages : uint8_t {
+ *     Sindarin    = (1 << 0),
+ *     Westron     = (1 << 1),
+ *     Rohirric    = (1 << 2),
+ *     BlackSpeech = (1 << 3)
+ * };
+ *
+ * my_code.cpp:
+ *
+ * #include "mbed_enum_flags.h"
+ * #include "external_enum.h"
+ *
+ * MBED_ENUM_FLAG_APPLY_OPERATORS(SpokenLanguages)
+ *
+ * SpokenLanguages gandalf = SpokenLanguages::Sindarin | SpokenLanguages::Westron |
+ *     SpokenLanguages::Rohirric | SpokenLanguages::BlackSpeech;
+ * @endcode
+ *
+ */
+#define MBED_ENUM_FLAG_APPLY_OPERATORS(T) \
+    inline constexpr T operator ~( T lhs) \
+    { \
+        return ( T ) (~static_cast<std::underlying_type_t< T > >(lhs)) ; \
+    } \
+    ENUM_FLAG_OPERATOR(T, |) \
+    ENUM_FLAG_OPERATOR(T, ^) \
+    ENUM_FLAG_OPERATOR(T, &) \
+    ENUM_FLAG_OPERATOR_LHS_REFERNCE(T, |=) \
```

Comment:
`REFERENCE`. Slightly indirect name though? The reference isn't the point. From standard terminology they could bee"binary operator" and "compound assignment operator".

Reply:
Ok, with this one, I was thinking of it not that in such a way as it only applied to enum classes that we wanted to make into bitflags. If I wanted to define the addition operator or the comma operator, I could do that with this same macro, as the signatures (with the exception of the exact symbol) are identical, and the name would still fit.

So in this case, I had `ENUM_FLAG` as a prefix, and `OPERATOR` and `OPERATOR_LHS_REFERENCE` as suffixes. But I do not mind narrowing down their usage, as this really is an internal macro anyway.

Reply:
Okay, I see how you were thinking. Not that bothered :)